File: sdks/typescript/src/resource_path.ts
// Resource-bound authority — the resource_path constraint's path model,
// segment-boundary matching, and the extension-constraint params value model
// (SPEC §5.7.1, §5.7.3). Mirrors the Go reference resource_path.go exactly.
//
// Paths are absolute logical POSIX-style paths: they name a location inside
// the resource's own namespace, not filesystem paths on any machine. The
// verifier compares bytes exactly — no Unicode normalization, no percent
// decoding, no case folding. Issuers and adapters pre-normalize (NFC) and
// pre-decode BEFORE constructing constraints and verifier context; nothing
// transforms a path after verification.

import {
  MAX_IDENTIFIER_LENGTH_BYTES,
  MAX_JSON_NESTING_DEPTH,
  type Constraint,
  type ConstraintType,
} from "./types.js";

// 2^53 - 1: the largest integer JavaScript (and JSON's number domain)
// represents exactly. params integers beyond this travel as decimal strings.
const MAX_SAFE_INTEGER_VALUE = Number.MAX_SAFE_INTEGER;

const _utf8 = new TextEncoder();

/** UTF-8 byte length of a string — Go's len() over a string. */
export function utf8ByteLength(s: string): number {
  return _utf8.encode(s).length;
}

/**
 * Validate p against the logical path model (SPEC §5.7.3) and return its
 * comparison form: at most one trailing "/" trimmed, except for the root
 * path "/" itself. Throws when p is not a valid logical path:
 *
 *   - must begin with "/" ("/" is the only separator)
 *   - no NUL byte, no backslash (rejected, never translated)
 *   - no "." or ".." segment (rejected outright, never resolved)
 *   - no empty interior segment ("/a//b" is invalid; "/" is valid)
 *
 * "%" is an ordinary literal byte: the path model has no percent-encoding.
 */
export function normalizeResourcePath(p: string): string {
  if (p === "") throw new Error("path is empty");
  if (!p.startsWith("/")) throw new Error("path must begin with '/'");
  if (p.indexOf("\u0000") >= 0) throw new Error("path contains a NUL byte");
  if (p.indexOf("\\") >= 0) {
    throw new Error("path contains a backslash; '/' is the only separator");
  }
  if (p === "/") return "/";
  // Trim at most one trailing slash before segment validation: "/docs/" ≡
  // "/docs", while "/docs//" leaves an empty interior segment and is invalid.
  const trimmed = p.endsWith("/") ? p.slice(0, -1) : p;
  const segments = trimmed.slice(1).split("/");
  for (const seg of segments) {
    if (seg === "") throw new Error("path contains an empty segment");
    if (seg === "." || seg === "..") {
      throw new Error(
        "path contains a dot-segment; dot-segments are rejected, never resolved",
      );
    }
  }
  return trimmed;
}

/**
 * Reports whether requestedPath is at or below pathPrefix under
 * segment-boundary prefix matching (SPEC §5.7.3). NOT glob matching and NOT
 * plain string-prefix matching: "/src" matches "/src" and "/src/a.ts", never
 * "/src-old/a.ts" or "/srcx". Either input being invalid yields false.
 */
export function resourcePathMatches(
  pathPrefix: string,
  requestedPath: string,
): boolean {
  let prefix: string;
  let path: string;
  try {
    prefix = normalizeResourcePath(pathPrefix);
  } catch {
    return false;
  }
  try {
    path = normalizeResourcePath(requestedPath);
  } catch {
    return false;
  }
  if (prefix === "/") return true; // root prefix matches every valid path
  if (path === prefix) return true;
  return path.startsWith(prefix + "/");
}

// Reports whether two valid prefixes are comparable under the prefix relation
// (one at or below the other). An absent prefix ("") orders as "/".
function resourcePathsNested(a: string, b: string): boolean {
  const aa = a === "" ? "/" : a;
  const bb = b === "" ? "/" : b;
  return resourcePathMatches(aa, bb) || resourcePathMatches(bb, aa);
}

/**
 * Enforce the issuance rule of SPEC §5.7.3: a certificate's resource_path
 * constraints must be jointly satisfiable. Rejects constraint sets naming
 * different resource_ids, and same-resource sets whose prefixes do not form a
 * nested chain. Individual constraints are also validated (non-empty
 * resource_id within MAX_IDENTIFIER_LENGTH_BYTES; well-formed path_prefix).
 * Throws on violation.
 *
 * Decoders do NOT call this — wire compatibility is not conditioned on
 * issuance hygiene; verification denies unsatisfiable sets fail-closed.
 */
export function validateResourceConstraints(constraints: Constraint[]): void {
  let resourceID = "";
  const prefixes: string[] = [];
  for (let i = 0; i < constraints.length; i++) {
    const c = constraints[i]!;
    if (c.type !== "resource_path") continue;
    const rid = c.resource_id ?? "";
    if (rid === "") {
      throw new Error(`constraint[${i}]: resource_path requires a non-empty resource_id`);
    }
    const ridLen = utf8ByteLength(rid);
    if (ridLen > MAX_IDENTIFIER_LENGTH_BYTES) {
      throw new Error(
        `constraint[${i}]: resource_id is ${ridLen} bytes, exceeding MAX_IDENTIFIER_LENGTH_BYTES (${MAX_IDENTIFIER_LENGTH_BYTES})`,
      );
    }
    const pp = c.path_prefix ?? "";
    if (pp !== "") {
      try {
        normalizeResourcePath(pp);
      } catch (e) {
        throw new Error(`constraint[${i}]: invalid path_prefix: ${(e as Error).message}`);
      }
    }
    if (resourceID === "") {
      resourceID = rid;
    } else if (rid !== resourceID) {
      throw new Error(
        `constraint[${i}]: certificate carries resource_path constraints naming different resource_ids — jointly unsatisfiable (SPEC §5.7.3); issue separate certificates`,
      );
    }
    for (const prev of prefixes) {
      if (!resourcePathsNested(prev, pp)) {
        throw new Error(
          `constraint[${i}]: same-resource path_prefix values "${prev}" and "${pp}" do not nest — jointly unsatisfiable (SPEC §5.7.3); narrow to one prefix or issue separate certificates`,
        );
      }
    }
    prefixes.push(pp);
  }
}

/**
 * Enforce the extension-constraint params value model (SPEC §5.7.1): null,
 * booleans, strings, safe integers (|n| ≤ 2^53−1), and arrays/objects of
 * these. Floats and non-integer numbers are prohibited; integers beyond the
 * safe range travel as decimal strings. `depth` is the container nesting
 * already consumed; nesting beyond MAX_JSON_NESTING_DEPTH is rejected. Throws
 * on violation.
 */
export function validateParamsValue(v: unknown, depth: number): void {
  if (depth > MAX_JSON_NESTING_DEPTH) {
    throw new Error(`params nesting exceeds MAX_JSON_NESTING_DEPTH (${MAX_JSON_NESTING_DEPTH})`);
  }
  if (v === null || typeof v === "boolean" || typeof v === "string") return;
  if (typeof v === "number") {
    if (!Number.isFinite(v) || !Number.isInteger(v)) {
      throw new Error("params values must be safe integers, not floats (carry non-integers as strings)");
    }
    if (Math.abs(v) > MAX_SAFE_INTEGER_VALUE) {
      throw new Error("params integer exceeds the safe-integer range; carry it as a decimal string");
    }
    return;
  }
  if (typeof v === "bigint") {
    if (v > BigInt(MAX_SAFE_INTEGER_VALUE) || v < -BigInt(MAX_SAFE_INTEGER_VALUE)) {
      throw new Error("params integer exceeds the safe-integer range; carry it as a decimal string");
    }
    return;
  }
  // Raw bytes are outside the value model (carried as base64 strings if at all).
  if (v instanceof Uint8Array) {
    throw new Error("params value of type Uint8Array is outside the restricted value model");
  }
  if (Array.isArray(v)) {
    for (const item of v) validateParamsValue(item, depth + 1);
    return;
  }
  if (typeof v === "object") {
    for (const item of Object.values(v as Record<string, unknown>)) {
      validateParamsValue(item, depth + 1);
    }
    return;
  }
  throw new Error(`params value of type ${typeof v} is outside the restricted value model`);
}

// The canonical v1 constraint kinds (SPEC §5.7.2). params is permitted only
// on constraint types NOT in this set.
const CANONICAL_CONSTRAINT_TYPES: ReadonlySet<string> = new Set([
  "geo_circle",
  "geo_polygon",
  "geo_bbox",
  "time_window",
  "max_speed_mps",
  "max_amount",
  "max_rate",
  "resource_path",
]);

/** Reports whether t is one of the canonical v1 constraint kinds. */
export function isCanonicalConstraintType(t: ConstraintType | string): boolean {
  return CANONICAL_CONSTRAINT_TYPES.has(t);
}

File: sdks/rust/src/resource_path.rs
//! Resource-bound authority — the `resource_path` constraint's path model,
//! segment-boundary matching, and the extension-constraint params value model
//! (SPEC §5.7.1, §5.7.3). Mirrors Go's `resource_path.go` byte-for-byte.
//!
//! Paths are absolute logical POSIX-style paths: they name a location inside
//! the resource's own namespace, not filesystem paths on any machine. The
//! verifier compares bytes exactly — no Unicode normalization, no percent
//! decoding, no case folding. Issuers and adapters pre-normalize (NFC) and
//! pre-decode BEFORE constructing constraints and verifier context; nothing
//! transforms a path after verification.

#[cfg(not(feature = "std"))]
use alloc::{format, string::String, string::ToString, vec::Vec};

use crate::canonical::wire_int::MAX_SAFE_INTEGER;
use crate::types::{
    is_canonical_constraint_type, Constraint, ParamsValue, MAX_IDENTIFIER_LENGTH_BYTES,
    MAX_JSON_NESTING_DEPTH,
};

/// Validate `p` against the logical path model (SPEC §5.7.3) and return its
/// comparison form: at most one trailing `/` trimmed, except for the root
/// path `/` itself. Returns `Err` when `p` is not a valid logical path:
///
///   - must begin with `/` (`/` is the only separator)
///   - no NUL byte, no backslash (rejected, never translated)
///   - no `.` or `..` segment (rejected outright, never resolved)
///   - no empty interior segment (`/a//b` is invalid; `/` is valid)
///
/// `%` is an ordinary literal byte: the path model has no percent-encoding.
pub fn normalize_resource_path(p: &str) -> Result<String, String> {
    if p.is_empty() {
        return Err("path is empty".to_string());
    }
    if !p.starts_with('/') {
        return Err("path must begin with '/'".to_string());
    }
    if p.as_bytes().contains(&0x00) {
        return Err("path contains a NUL byte".to_string());
    }
    if p.contains('\\') {
        return Err("path contains a backslash; '/' is the only separator".to_string());
    }
    if p == "/" {
        return Ok("/".to_string());
    }
    // Trim at most one trailing slash before segment validation: "/docs/" ≡
    // "/docs", while "/docs//" leaves an empty interior segment and is invalid.
    let trimmed = p.strip_suffix('/').unwrap_or(p);
    for seg in trimmed[1..].split('/') {
        match seg {
            "" => return Err("path contains an empty segment".to_string()),
            "." | ".." => {
                return Err(
                    "path contains a dot-segment; dot-segments are rejected, never resolved"
                        .to_string(),
                )
            }
            _ => {}
        }
    }
    Ok(trimmed.to_string())
}

/// Report whether `requested_path` is at or below `path_prefix` under
/// segment-boundary prefix matching (SPEC §5.7.3). This is NOT glob matching
/// and NOT plain string-prefix matching: `/src` matches `/src` and
/// `/src/a.ts`, never `/src-old/a.ts` or `/srcx`.
pub fn resource_path_matches(path_prefix: &str, requested_path: &str) -> bool {
    let prefix = match normalize_resource_path(path_prefix) {
        Ok(p) => p,
        Err(_) => return false,
    };
    let path = match normalize_resource_path(requested_path) {
        Ok(p) => p,
        Err(_) => return false,
    };
    if prefix == "/" {
        // The root prefix matches every valid path.
        return true;
    }
    if path == prefix {
        return true;
    }
    path.starts_with(&format!("{}/", prefix))
}

/// Report whether two prefixes are comparable under the prefix relation (one
/// at or below the other). An absent prefix orders as `/` (SPEC §5.7.3).
fn resource_paths_nested(a: &str, b: &str) -> bool {
    let a = if a.is_empty() { "/" } else { a };
    let b = if b.is_empty() { "/" } else { b };
    resource_path_matches(a, b) || resource_path_matches(b, a)
}

/// Enforce the issuance rule of SPEC §5.7.3: a certificate's `resource_path`
/// constraints must be jointly satisfiable. Rejects constraint sets naming
/// different `resource_id`s, and same-resource sets whose prefixes do not
/// form a nested chain under the prefix relation. Individual constraints are
/// also validated (non-empty `resource_id` within `MAX_IDENTIFIER_LENGTH_BYTES`;
/// well-formed `path_prefix`).
///
/// Decoders do NOT call this — wire compatibility is not conditioned on
/// issuance hygiene; verification denies unsatisfiable sets fail-closed.
pub fn validate_resource_constraints(constraints: &[Constraint]) -> Result<(), String> {
    let mut resource_id: Option<&str> = None;
    let mut prefixes: Vec<&str> = Vec::new();
    for (i, c) in constraints.iter().enumerate() {
        if c.kind != "resource_path" {
            continue;
        }
        if c.resource_id.is_empty() {
            return Err(format!(
                "constraint[{}]: resource_path requires a non-empty resource_id",
                i
            ));
        }
        if c.resource_id.len() > MAX_IDENTIFIER_LENGTH_BYTES {
            return Err(format!(
                "constraint[{}]: resource_id is {} bytes, exceeding MAX_IDENTIFIER_LENGTH_BYTES ({})",
                i,
                c.resource_id.len(),
                MAX_IDENTIFIER_LENGTH_BYTES
            ));
        }
        let this_prefix = c.path_prefix.as_deref().unwrap_or("");
        if !this_prefix.is_empty() {
            if let Err(e) = normalize_resource_path(this_prefix) {
                return Err(format!("constraint[{}]: invalid path_prefix: {}", i, e));
            }
        }
        match resource_id {
            None => resource_id = Some(&c.resource_id),
            Some(rid) if c.resource_id != rid => {
                return Err(format!(
                    "constraint[{}]: certificate carries resource_path constraints naming different resource_ids — jointly unsatisfiable (SPEC §5.7.3); issue separate certificates",
                    i
                ));
            }
            _ => {}
        }
        for prev in &prefixes {
            if !resource_paths_nested(prev, this_prefix) {
                return Err(format!(
                    "constraint[{}]: same-resource path_prefix values \"{}\" and \"{}\" do not nest — jointly unsatisfiable (SPEC §5.7.3); narrow to one prefix or issue separate certificates",
                    i, prev, this_prefix
                ));
            }
        }
        prefixes.push(this_prefix);
    }
    Ok(())
}

/// Enforce the extension-constraint params value model (SPEC §5.7.1): null,
/// booleans, strings, safe integers (|n| ≤ 2^53−1), and arrays/objects of
/// these. `depth` is the container nesting already consumed; nesting beyond
/// `MAX_JSON_NESTING_DEPTH` is rejected.
///
/// The [`ParamsValue`] type cannot represent floats or raw bytes, so the
/// remaining checks are the safe-integer bound and the nesting depth.
pub fn validate_params_value(v: &ParamsValue, depth: usize) -> Result<(), String> {
    if depth > MAX_JSON_NESTING_DEPTH {
        return Err(format!(
            "params nesting exceeds MAX_JSON_NESTING_DEPTH ({})",
            MAX_JSON_NESTING_DEPTH
        ));
    }
    match v {
        ParamsValue::Null | ParamsValue::Bool(_) | ParamsValue::Str(_) => Ok(()),
        ParamsValue::Int(n) => {
            if *n > MAX_SAFE_INTEGER || *n < -MAX_SAFE_INTEGER {
                return Err(
                    "params integer exceeds the safe-integer range; carry it as a decimal string"
                        .to_string(),
                );
            }
            Ok(())
        }
        ParamsValue::Array(items) => {
            for item in items {
                validate_params_value(item, depth + 1)?;
            }
            Ok(())
        }
        ParamsValue::Object(map) => {
            for item in map.values() {
                validate_params_value(item, depth + 1)?;
            }
            Ok(())
        }
    }
}

/// Validate a constraint's `params` field against the value model and the
/// canonical-type restriction (SPEC §5.7.1). Shared by issuance and decode.
pub fn validate_constraint_params(c: &Constraint) -> Result<(), String> {
    if let Some(params) = &c.params {
        if is_canonical_constraint_type(&c.kind) {
            return Err(format!(
                "canonical constraint type \"{}\" must not carry params",
                c.kind
            ));
        }
        // The params object is the depth-0 container (Go: ValidateParamsValue
        // on the map iterates its values at depth+1), so values start at 1.
        for value in params.values() {
            validate_params_value(value, 1)?;
        }
    }
    Ok(())
}
